hs_core/models.py

```diff
@@ -1508,41 +1512,205 @@ class Meta:
+                relpath.startswith(self.content_object.resource_federation_path + '/'):
+            if not storage.exists(path):
+                raise ValidationError("Federated path does not exist in irods")
+            plen = len(self.content_object.resource_federation_path)
+            relpath = relpath[plen+1:]  # omit /
+
+            # strip resource id from path
+            if relpath.startswith(file_path):
+                plen = len(file_path)
+                relpath = relpath[plen:]  # omit /
+            else:
+                raise ValidationError("Malformed federated resource path")
+        elif path.startswith(file_path):
+            # strip optional local path prefix
+            if not storage.exists(path):
+                raise ValidationError("Federated path does not exist in irods")
+            plen = len(file_path)
+            relpath = relpath[plen:]  # strip local prefix, omit /
+
+        # now we have folder/file. We could have gotten this from the input, or
+        # from stripping qualification folders. Note that this can contain
+        # misnamed header content misinterpreted as a folder
+        if '/' in relpath:
+            folder, base = relpath.rsplit('/', 1)
+            abspath = get_path(self, base, folder=folder)
+            print("abspath = " + abspath)
+            if not storage.exists(abspath):
+                raise ValidationError("Local path does not exist in irods")
+        else:
+            folder = None
+            base = relpath
+            abspath = get_path(self, base, folder=folder)
+            print("abspath = " + abspath)
+            if not storage.exists(abspath):
+                raise ValidationError("Local path does not exist in irods")
+
+        return folder, base
+
+    # def rename(self, new_name):
+    #     """ rename a file, setting all path variables appropriately """
+    #     pass
+
+    # def copy_irods(self, source_path, dest_path=None):
+    #     """ copy an irods file into this FileField, setting all paths appropriately """
+    #     pass
+
+    # def move_irods(self, source_path, dest_path=None):
+    #     """ move an irods file into this object, setting all paths appropriately """
+
+    # classmethods do things that affect all files.
+
+    @classmethod
+    def create_folder(cls, resource, folder):
+        """ create a folder for a resource """
+        # avoid import loop
+        from hs_core.views.utils import create_folder
+        _path_is_allowed(folder)
+        # TODO: move code from location used below to here
+        create_folder(resource.short_id, folder)
+
+    @classmethod
+    def remove_folder(cls, resource, folder, user):
+        """ remove a folder for a resource """
+        # avoid import loop
+        from hs_core.views.utils import remove_folder
+        _path_is_allowed(folder)
+        # TODO: move code from location used below to here
+        remove_folder(user, resource.short_id, folder)
```

Comment:
I know logically it sounds more reasonable to put file operations in ResourceFile class, however, our ResourceFile class definition has a many-to-one foreign key relationship with BaseResource class, so one resource contains multiple ResourceFile objects with each ResourceFile object corresponding to each single resource file. ```create_folder``` and ```remove_folder``` is resource level operations since it is not related to a single contained resource file instance. So I think folder-level operations belong to BaseResource class methods or as external utility methods.